lib/checkstl.cpp

```diff
@@ -229,10 +264,52 @@ static std::string getContainerName(const Token *containerToken)
     return ret;
 }
 
+static const Token* findIteratorContainer(const Token* start, const Token* end, unsigned int id)
+{
+    const Token* containerToken = nullptr;
+    for(const Token* tok = start; tok != end; tok = tok->next())
+    {
+        if (Token::Match(tok, "%varid% = %name% . %name% (", id))
+        {
+            if (tok->tokAt(5)->valueType() && tok->tokAt(5)->valueType()->type == ValueType::Type::ITERATOR)
+            {
+                containerToken = tok->tokAt(2);
+            }
+        }
+        else if (Token::Match(tok, "%varid% = %name% (", id))
+        {
+            // Prevent FP: iterator is assigned to something
+            // TODO: Fix it in future
+            containerToken = nullptr;
+        }
+    }
+    return containerToken;
+}
+
+enum OperandPosition
+{
+    None,
+    Left,
+    Right
+};
+
 void CheckStl::iterators()
 {
     const SymbolDatabase *symbolDatabase = mTokenizer->getSymbolDatabase();
 
+    // Filling map of iterators id and their scope begin
+    std::map<unsigned int, const Token*> iteratorScopeBegin;
+    for (const Variable* var : symbolDatabase->variableList()) {
+        bool inconclusiveType=false;
+        if (!isIterator(var, inconclusiveType))
+            continue;
+        const unsigned int iteratorId = var->declarationId();
+        if (iteratorId != 0)
+            iteratorScopeBegin[iteratorId] = var->nameToken();
+    }
+    // Storage to save found comparison problems to avoid duplicate error messages
+    std::set<const Token*> foundOperatorErrors;
```

Comment:
@danmar I know about the `reportErr` function, I have seen its code and `mErrorList` is checked before writting log message. On the other hand, I have cases when two iterators are checked:
```
void foo()
{
    std::list<int> l1;
    std::list<int> l2;
    std::list<int>::iterator it1 = l1.end();
    std::list<int>::iterator it2 = l2.begin();
    if (it1 != it2)
    {    
    }
}
```

and we decided (after first code review) to point to containers when write error message and as a result we have the following situation, for instance:
```
[test.cpp:7] -> [test.cpp:6] -> [test.cpp:5]: (error) Comparison of iterators from containers 'l1' and 'l2'.\n
[test.cpp:7] -> [test.cpp:5] -> [test.cpp:6]: (error) Comparison of iterators from containers 'l1' and 'l2'.\n
```
Because my implemetation will find the problem for `it1` and also find the same problem for `it2`. Without pointing to contaners it would be suppresed by `CppCheck::reportErr`. For avoiding it I introduced `foundOperatorErrors` set. Am I right?

Reply:
ok I see. keep this code then.